File: src/signatureHelp.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { createChecker } from "./checker.ts";
import { loadJdkStub } from "./jdkStub.ts";
import { getNodeAtPosition } from "./nodeAtPosition.ts";
import { createProgram } from "./program.ts";
import { type CallExpression, type Node, SyntaxKind } from "./types.ts";
import { type Uri } from "./workspace.ts";

// The checker-side pieces signature help builds on: resolveCallCandidates (every
// overload a call could bind to), parameterLabelsOf, and signatureOfDeclaration.
// The server handler is a thin position-to-call wrapper over these.

function callAtMarker(text: string, marker = "/*|*/") {
  const offset = text.indexOf(marker);
  const clean = text.replace(marker, "");
  const program = createProgram();
  loadJdkStub(program);
  program.setOpenDocument("file:///T.java" as Uri, clean, 1);
  const checker = createChecker(program);
  const sf = program.getSourceFile("file:///T.java" as Uri)!;
  // Mirrors the server's callAt, including the retry one position left for a
  // cursor sitting at the recovered end of an unclosed call.
  for (const at of [offset, offset - 1]) {
    let node: Node | undefined = getNodeAtPosition(sf, at);
    for (; node; node = node.parent) {
      if (node.kind !== SyntaxKind.CallExpression) continue;
      const call = node as CallExpression;
      if (offset > call.expression.end && offset <= call.end) return { call, checker };
    }
  }
  throw new Error("no call at marker");
}

test("call candidates list every overload", () => {
  const { call, checker } = callAtMarker(
    "class C { int f(int x){return x;} int f(String s){return 0;} void m(){ f(/*|*/ } }",
  );
  const sigs = checker.resolveCallCandidates(call).map(d => checker.signatureOfDeclaration(d));
  expect(sigs.sort()).toEqual(["int f(String s)", "int f(int x)"]);
});

test("parameter labels are the written parameter texts", () => {
  const { call, checker } = callAtMarker(
    "class C { int add(int first, long second){return 0;} void m(){ add(1,/*|*/ } }",
  );
  const decl = checker.resolveCall(call)!;
  expect(checker.parameterLabelsOf(decl)).toEqual(["int first", "long second"]);
});

test("candidates resolve through a receiver, including inherited overloads", () => {
  const { call, checker } = callAtMarker(
    "class B { void g(int x){} } class D extends B { void g(String s){} void m(){ this.g(/*|*/ } }",
  );
  const labels = checker
    .resolveCallCandidates(call)
    .map(d => checker.signatureOfDeclaration(d))
    .sort();
  expect(labels).toEqual(["void g(String s)", "void g(int x)"]);
});

test("the chosen overload matches the argument types", () => {
  const { call, checker } = callAtMarker(
    'class C { int f(int x){return x;} int f(String s){return 0;} void m(){ f("a"/*|*/) } }',
  );
  const resolved = checker.resolveCall(call)!;
  expect(checker.signatureOfDeclaration(resolved)).toBe("int f(String s)");
});

File: togo/internal/services/signature_help_test.go
package services

import (
	"sort"
	"strings"
	"testing"

	"github.com/nikeee/cappu/internal/compiler"
)

// Port of src/services/signatureHelp.test.ts. Signature help is the checker's
// resolveCallCandidates / parameterLabelsOf / signatureOfDeclaration over the
// call under the cursor; the server handler is a thin position-to-call wrapper.

func callAtMarker(t *testing.T, text string) (*compiler.Node, *compiler.Checker) {
	t.Helper()
	const marker = "/*|*/"
	offset := strings.Index(text, marker)
	clean := strings.Replace(text, marker, "", 1)
	program := compiler.NewProgram()
	compiler.LoadJdkStub(program)
	program.SetOpenDocument("file:///T.java", clean, 1)
	checker := compiler.NewChecker(program)
	sf := program.GetSourceFile("file:///T.java")
	for _, at := range []int{offset, offset - 1} {
		node := compiler.GetNodeAtPosition(sf, at)
		for ; node != nil; node = node.Parent {
			if node.Kind != compiler.CallExpression {
				continue
			}
			call := node.AsCallExpression()
			if offset > call.Expression.End && offset <= node.End {
				return node, checker
			}
		}
	}
	t.Fatal("no call at marker")
	return nil, nil
}

func TestCallCandidatesListOverloads(t *testing.T) {
	call, checker := callAtMarker(t, "class C { int f(int x){return x;} int f(String s){return 0;} void m(){ f(/*|*/ } }")
	var sigs []string
	for _, d := range checker.ResolveCallCandidates(call) {
		s, _ := checker.SignatureOfDeclaration(d)
		sigs = append(sigs, s)
	}
	sort.Strings(sigs)
	if len(sigs) != 2 || sigs[0] != "int f(String s)" || sigs[1] != "int f(int x)" {
		t.Errorf("candidates = %v", sigs)
	}
}

func TestParameterLabelsWrittenText(t *testing.T) {
	call, checker := callAtMarker(t, "class C { int add(int first, long second){return 0;} void m(){ add(1,/*|*/ } }")
	decl := checker.ResolveCall(call)
	labels := checker.ParameterLabelsOf(decl)
	if len(labels) != 2 || labels[0] != "int first" || labels[1] != "long second" {
		t.Errorf("labels = %v", labels)
	}
}

func TestCandidatesThroughReceiverInherited(t *testing.T) {
	call, checker := callAtMarker(t, "class B { void g(int x){} } class D extends B { void g(String s){} void m(){ this.g(/*|*/ } }")
	var labels []string
	for _, d := range checker.ResolveCallCandidates(call) {
		s, _ := checker.SignatureOfDeclaration(d)
		labels = append(labels, s)
	}
	sort.Strings(labels)
	if len(labels) != 2 || labels[0] != "void g(String s)" || labels[1] != "void g(int x)" {
		t.Errorf("candidates = %v", labels)
	}
}

func TestChosenOverloadMatchesArgs(t *testing.T) {
	call, checker := callAtMarker(t, "class C { int f(int x){return x;} int f(String s){return 0;} void m(){ f(\"a\"/*|*/) } }")
	resolved := checker.ResolveCall(call)
	if sig, _ := checker.SignatureOfDeclaration(resolved); sig != "int f(String s)" {
		t.Errorf("resolved = %q, want int f(String s)", sig)
	}
}
